{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3947b1a5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-11T11:23:54.533065Z",
     "start_time": "2022-11-11T11:23:54.529093Z"
    }
   },
   "outputs": [],
   "source": [
    "#########DeepMRMP##########\n",
    "######DeepMRMP: A new predictor for multiple types of RNA modification sites using deep learning#############################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "35980531",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-11T11:40:08.840791Z",
     "start_time": "2022-11-11T11:40:08.708623Z"
    }
   },
   "outputs": [],
   "source": [
    "##单核苷酸频率，仅四个结果\n",
    "def nc(seq, seq_length):\n",
    "    A = round(seq.count('A') / seq_length, 3)\n",
    "    U = round(seq.count('U') / seq_length, 3)\n",
    "    C = round(seq.count('C') / seq_length, 3)\n",
    "    G = round(seq.count('G') / seq_length, 3)\n",
    "    return A, U, C, G\n",
    "\n",
    "#双核苷酸频率，16个结果\n",
    "def dnc(seq, seq_length):\n",
    "    AA = round(seq.count('AA') / seq_length, 3)\n",
    "    AU = round(seq.count('AU') / seq_length, 3)\n",
    "    AC = round(seq.count('AC') / seq_length, 3)\n",
    "    AG = round(seq.count('AG') / seq_length, 3)\n",
    "    UA = round(seq.count('UA') / seq_length, 3)\n",
    "    UU = round(seq.count('UU') / seq_length, 3)\n",
    "    UC = round(seq.count('UC') / seq_length, 3)\n",
    "    UG = round(seq.count('UG') / seq_length, 3)\n",
    "    CA = round(seq.count('CA') / seq_length, 3)\n",
    "    CU = round(seq.count('CU') / seq_length, 3)\n",
    "    CC = round(seq.count('CC') / seq_length, 3)\n",
    "    CG = round(seq.count('CG') / seq_length, 3)\n",
    "    GA = round(seq.count('GA') / seq_length, 3)\n",
    "    GU = round(seq.count('GU') / seq_length, 3)\n",
    "    GC = round(seq.count('GC') / seq_length, 3)\n",
    "    GG = round(seq.count('GG') / seq_length, 3)\n",
    "    return AA, AU, AC, AG, UA, UU, UC, UG, CA, CU, CC, CG, GA, GU, GC, GG\n",
    "\n",
    "#三核苷酸频率，64个结果\n",
    "def tnc(seq, seq_length):\n",
    "    RNA = ['A', 'U', 'C', 'G']\n",
    "    tri_nucleotide_values = []\n",
    "    tri_nucleotide_dict = {\"\".join(i): 0 for i in product(RNA, repeat=3)}\n",
    "    for tri_nucleotide in tri_nucleotide_dict.keys():\n",
    "        tri_nucleotide_dict[tri_nucleotide] = round(seq.count(tri_nucleotide) / seq_length, 3)\n",
    "    for dict_value in tri_nucleotide_dict.values():\n",
    "        tri_nucleotide_values.append(dict_value)\n",
    "    return tri_nucleotide_values\n",
    "\n",
    "#四核苷酸频率，256个结果\n",
    "def tenc(seq, seq_length):\n",
    "    RNA = ['A', 'U', 'C', 'G']\n",
    "    fourth_nucleotide_values = []\n",
    "    fourth_nucleotide_dict = {\"\".join(i): 0 for i in product(RNA, repeat=4)}\n",
    "    for fourth_nucleotide in fourth_nucleotide_dict.keys():\n",
    "        fourth_nucleotide_dict[fourth_nucleotide] = round(seq.count(fourth_nucleotide) / seq_length, 3)\n",
    "    for dict_value in fourth_nucleotide_dict.values():\n",
    "        fourth_nucleotide_values.append(dict_value)\n",
    "    return fourth_nucleotide_values\n",
    "\n",
    "#五核苷酸频率，1024个结果\n",
    "def pnc(seq, seq_length):\n",
    "    RNA = ['A', 'U', 'C', 'G']\n",
    "    fifth_nucleotide_values = []\n",
    "    fifth_nucleotide_dict = {\"\".join(i): 0 for i in product(RNA, repeat=5)}\n",
    "    for fifth_nucleotide in fifth_nucleotide_dict.keys():\n",
    "        fifth_nucleotide_dict[fifth_nucleotide] = round(seq.count(fifth_nucleotide) / seq_length, 3)\n",
    "    for dict_value in fifth_nucleotide_dict.values():\n",
    "        fifth_nucleotide_values.append(dict_value)\n",
    "    return fifth_nucleotide_values\n",
    "\n",
    "#理化性质特征\n",
    "def ncp(seq, seq_length):\n",
    "    ncp_lsit = [None] * seq_length * 3\n",
    "    for j in range(seq_length):\n",
    "        if seq[j] == 'A':\n",
    "            ncp_lsit[j * 3] = 1\n",
    "            ncp_lsit[j * 3 + 1] = 1\n",
    "            ncp_lsit[j * 3 + 2] = 1\n",
    "        elif seq[j] == 'U':\n",
    "            ncp_lsit[j * 3] = 0\n",
    "            ncp_lsit[j * 3 + 1] = 0\n",
    "            ncp_lsit[j * 3 + 2] = 1\n",
    "        elif seq[j] == 'C':\n",
    "            ncp_lsit[j * 3] = 0\n",
    "            ncp_lsit[j * 3 + 1] = 1\n",
    "            ncp_lsit[j * 3 + 2] = 0\n",
    "        elif seq[j] == 'G':\n",
    "            ncp_lsit[j * 3] = 1\n",
    "            ncp_lsit[j * 3 + 1] = 0\n",
    "            ncp_lsit[j * 3 + 2] = 0\n",
    "    return ncp_lsit  #应该是理化性质，A：111  U：001  C：010  G：100、核苷酸化学性质\n",
    "\n",
    "#单核苷酸从头到尾的密度\n",
    "def nd(seq, seq_length):\n",
    "    nd_list = [None] * seq_length\n",
    "    for j in range(seq_length):\n",
    "        if seq[j] == 'A':\n",
    "            nd_list[j] = round(seq[0:j].count('A') / (j + 1), 3)\n",
    "        elif seq[j] == 'U':\n",
    "            nd_list[j] = round(seq[0:j].count('U') / (j + 1), 3)\n",
    "        elif seq[j] == 'C':\n",
    "            nd_list[j] = round(seq[0:j].count('C') / (j + 1), 3)\n",
    "        elif seq[j] == 'G':\n",
    "            nd_list[j] = round(seq[0:j].count('G') / (j + 1), 3)\n",
    "    return nd_list\n",
    "\n",
    "#双核苷酸从头到尾的密度\n",
    "def dnd(seq, seq_length):\n",
    "    nnd_list = [None] * (seq_length-1)\n",
    "    for j in range(seq_length-1):\n",
    "        if seq[j:j + 2] == 'AA':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('AA') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'AU':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('AU') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'AC':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('AC') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'AG':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('AG') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'UA':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('UA') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'UU':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('UU') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'UC':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('UC') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'UG':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('UG') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'CA':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('CA') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'CU':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('CU') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'CC':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('CC') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'CG':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('CG') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'GA':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('GA') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'GU':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('GU') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'GC':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('GC') / (j+1), 3)\n",
    "        elif seq[j:j + 2] == 'GG':\n",
    "            nnd_list[j] = round(seq[0:j+2].count('GG') / (j+1), 3)\n",
    "    return nnd_list\n",
    "\n",
    "#单热编码\n",
    "def onehot(seq, seq_length):\n",
    "    ot = [0, 0, 0, 0]\n",
    "    ONE_HOT = []\n",
    "    for i in seq:\n",
    "        if i == 'A':\n",
    "            ot = [1, 0, 0, 0]\n",
    "        elif i == 'U':\n",
    "            ot = [0, 1, 0, 0]\n",
    "        elif i == 'C':\n",
    "            ot = [0, 0, 1, 0]\n",
    "        elif i == 'G':\n",
    "            ot = [0, 0, 0, 1]\n",
    "        ONE_HOT = ONE_HOT + ot\n",
    "    return ONE_HOT\n",
    "\n",
    "#四个二进制数表示双核苷酸编码\n",
    "def dbe(seq, seq_length):\n",
    "    dot = [0, 0, 0, 0]\n",
    "    seq_length = len(seq)\n",
    "    dbe = []\n",
    "    for j in range(seq_length - 1):\n",
    "        if seq[j:j + 2] == 'AA':\n",
    "            dot = [0, 0, 0, 0]\n",
    "        elif seq[j:j + 2] == 'AU':\n",
    "            dot = [0, 0, 0, 1]\n",
    "        elif seq[j:j + 2] == 'AC':\n",
    "            dot = [0, 0, 1, 0]\n",
    "        elif seq[j:j + 2] == 'AG':\n",
    "            dot = [0, 0, 1, 1]\n",
    "        elif seq[j:j + 2] == 'UA':\n",
    "            dot = [0, 1, 0, 0]\n",
    "        elif seq[j:j + 2] == 'UU':\n",
    "            dot = [0, 1, 0, 1]\n",
    "        elif seq[j:j + 2] == 'UC':\n",
    "            dot = [0, 1, 1, 0]\n",
    "        elif seq[j:j + 2] == 'UG':\n",
    "            dot = [0, 1, 1, 1]\n",
    "        elif seq[j:j + 2] == 'CA':\n",
    "            dot = [1, 0, 0, 0]\n",
    "        elif seq[j:j + 2] == 'CU':\n",
    "            dot = [1, 0, 0, 1]\n",
    "        elif seq[j:j + 2] == 'CC':\n",
    "            dot = [1, 0, 1, 0]\n",
    "        elif seq[j:j + 2] == 'CG':\n",
    "            dot = [1, 0, 1, 1]\n",
    "        elif seq[j:j + 2] == 'GA':\n",
    "            dot = [1, 1, 0, 0]\n",
    "        elif seq[j:j + 2] == 'GU':\n",
    "            dot = [1, 1, 0, 1]\n",
    "        elif seq[j:j + 2] == 'GC':\n",
    "            dot = [1, 1, 1, 0]\n",
    "        elif seq[j:j + 2] == 'GG':\n",
    "            dot = [1, 1, 1, 1]\n",
    "        dbe = dbe + dot\n",
    "    return dbe\n",
    "\n",
    "#双核苷酸的11种理化性质编码\n",
    "def dpcp(seq, seq_length):\n",
    "    PC15DNC16 = []\n",
    "    Shift = [-0.08, -0.06, 0.23, -0.04, 0.07, 0.23, 0.07, -0.01, 0.11, -0.04, -0.01, 0.3, -0.02, -0.08, 0.07, 0.11]\n",
    "    Slide = [-1.27, -1.36, -1.43, -1.5, -1.7, -1.43, -1.39, -1.78, -1.46, -1.5, -1.78, -1.89, -1.45, -1.27, -1.7, -1.46]\n",
    "    Rise = [3.18, 3.24, 3.24, 3.3, 3.38, 3.24, 3.22, 3.32, 3.09, 3.3, 3.32, 3.3, 3.26, 3.18, 3.38, 3.09]\n",
    "    Tilt = [-0.8, 1.1, 0.8, 0.5, 1.3, 0.8, 0, 0.3, 1, 0.5, 0.3, -0.1, -0.2, -0.8, 1.3, 1]\n",
    "    Roll = [7, 7.1, 4.8, 8.5, 9.4, 4.8, 6.1, 12.1, 9.9, 8.5, 8.7, 12.1, 10.7, 7, 9.4, 9.9]\n",
    "    Twist = [31, 33, 32, 30, 32, 32, 35, 32, 31, 30, 32, 27, 32, 31, 32, 31]\n",
    "    Stacking_energy = [-13.7, -15.4, -13.8, -14, -14.2, -13.8, -16.9, -11.1, -14.4, -14, -11.1, -15.6, -16, -13.7, -14.2, -14.4]\n",
    "    Enthalpy = [-6.6, -5.7, -10.2, -7.6, -13.3, -10.2, -14.2, -12.2, -10.5, -7.6, -12.2, -8, -8.1, -6.6, -10.2, -7.6]\n",
    "    Entropy = [-18.4, -15.5, -26.2, -19.2, -35.5, -26.2, -34.9, -29.7, -27.8, -19.2, -29.7, -19.4, -22.6, -18.4, -26.2, -19.2]\n",
    "    Free_energy = [-0.93, -1.1, -2.24, -2.08, -2.35, -2.24, -3.42, -3.26, -2.11, -2.08, -3.26, -2.36, -1.33, -0.93, -2.35,-2.11]\n",
    "    Hydrophilicity = [0.04, 0.14, 0.14, 0.08, 0.1, 0.27, 0.26, 0.17, 0.21, 0.52, 0.49, 0.35, 0.21, 0.44, 0.48, 0.34]\n",
    "    name_list = [Shift, Slide, Rise, Tilt, Roll, Twist, Stacking_energy, Enthalpy, Entropy, Free_energy, Hydrophilicity]\n",
    "    for i in name_list:\n",
    "        AA = [round(seq.count('AA')*i[0] / seq_length, 3)]\n",
    "        AU = [round(seq.count('AU')*i[1] / seq_length, 3)]\n",
    "        AC = [round(seq.count('AC')*i[2] / seq_length, 3)]\n",
    "        AG = [round(seq.count('AG')*i[3] / seq_length, 3)]\n",
    "        UA = [round(seq.count('UA')*i[4] / seq_length, 3)]\n",
    "        UU = [round(seq.count('UU')*i[5] / seq_length, 3)]\n",
    "        UC = [round(seq.count('UC')*i[6] / seq_length, 3)]\n",
    "        UG = [round(seq.count('UG')*i[7] / seq_length, 3)]\n",
    "        CA = [round(seq.count('CA')*i[8] / seq_length, 3)]\n",
    "        CU = [round(seq.count('CU')*i[9] / seq_length, 3)]\n",
    "        CC = [round(seq.count('CC')*i[10] / seq_length, 3)]\n",
    "        CG = [round(seq.count('CG')*i[11] / seq_length, 3)]\n",
    "        GA = [round(seq.count('GA')*i[12] / seq_length, 3)]\n",
    "        GU = [round(seq.count('GU')*i[13] / seq_length, 3)]\n",
    "        GC = [round(seq.count('GC')*i[14] / seq_length, 3)]\n",
    "        GG = [round(seq.count('GG')*i[15] / seq_length, 3)]\n",
    "        PC15DNC16 = PC15DNC16 + AA + AU + AC + AG + UA + UU + UC + UG + CA + CU + CC + CG + GA + GU + GC +GG\n",
    "    return PC15DNC16  #二核苷酸理化性质\n",
    "\n",
    "#单热编码\n",
    "def onehotkey(seq, tag):\n",
    "    \"\"\"\n",
    "    one hot coding\n",
    "    :param seq:\n",
    "    :param tag:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    tag = np.array(tag)\n",
    "    for num in range(len(seq)):\n",
    "        seq[num] = seq[num].strip('\\n')\n",
    "\n",
    "    letterDict = {}\n",
    "    letterDict[\"A\"] = 0\n",
    "    letterDict[\"C\"] = 1\n",
    "    letterDict[\"G\"] = 2\n",
    "    letterDict[\"U\"] = 3\n",
    "    letterDict[\"T\"] = 3\n",
    "    CategoryLen = 4\n",
    "    probMatr = np.zeros((len(seq), 1, len(seq[0]), CategoryLen))\n",
    "    sampleNo = 0\n",
    "\n",
    "    for sequence in seq:\n",
    "        RNANo = 0\n",
    "        for RNA in sequence:\n",
    "            try:\n",
    "                index = letterDict[RNA]\n",
    "                probMatr[sampleNo][0][RNANo][index] = 1\n",
    "                RNANo += 1\n",
    "            except:\n",
    "                RNANo += 1\n",
    "        sampleNo += 1\n",
    "    return probMatr, tag\n",
    "\n",
    "#双核苷酸性质编码，10种\n",
    "def AAindexVector(shuffled_SEQ, shuffled_TAG):\n",
    "    \"\"\"add T\"\"\"\n",
    "    tag = np.array(shuffled_TAG)\n",
    "\n",
    "    for num in range(len(shuffled_SEQ)):\n",
    "        shuffled_SEQ[num] = shuffled_SEQ[num].strip('\\n')\n",
    "\n",
    "    letterDict = {}\n",
    "    letterDict[\"AA\"] = [-0.83,  -0.67,  -1.13,  1.34,   -1.84,  -0.26,  1.72,   0.95,   0.29,   1.35]\n",
    "    letterDict[\"AC\"] = [-0.14,  -1.64,  1.50,   0.49,   0.54,   0.34,   -0.27,  -0.31,  0.23,   -0.26]\n",
    "    letterDict[\"AG\"] = [0.55,   0.00,   -0.79,  0.12,   0.09,   -0.86,  0.13,   0.82,   0.10,   0.28]\n",
    "    letterDict[\"AU\"] = [-0.14,  -0.62,  -0.96,  0.87,   0.98,   0.93,   0.61,   1.42,   -0.83,  1.35]\n",
    "    letterDict[\"AT\"] = [-0.14,  -0.62,  -0.96,  0.87,   0.98,   0.93,   0.61,   1.42,   -0.83,  1.35]\n",
    "    letterDict[\"CA\"] = [-1.87,  0.62,   0.48,   0.33,   0.83,   -0.26,  0.15,   -0.57,  -0.17,  0.14]\n",
    "    letterDict[\"CC\"] = [0.78,   0.09,   -0.53,  -1.36,  -0.20,  0.34,   -1.13,  -0.88,  2.02,   -1.34]\n",
    "    letterDict[\"CG\"] = [0.55,   1.60,   2.09,   -1.95,  -0.80,  -2.65,  0.06,   0.79,   -0.97,  -0.13]\n",
    "    letterDict[\"CU\"] = [0.55,   0.00,   -0.79,  0.12,   0.09,   -0.86,  0.13,   0.82,   0.10,   0.28]\n",
    "    letterDict[\"CT\"] = [0.55,   0.00,   -0.79,  0.12,   0.09,   -0.86,  0.13,   0.82,   0.10,   0.28]\n",
    "    letterDict[\"GA\"] = [1.47,   0.40,   0.14,   -0.94,  1.28,   0.34,   -0.72,  -1.82,  -0.04,  -0.53]\n",
    "    letterDict[\"GC\"] = [-0.37,  -1.07,  0.14,   0.71,   -0.65,  2.13,   -1.77,  -1.72,  -1.83,  -2.01]\n",
    "    letterDict[\"GG\"] = [0.78,   1.60,   -0.53,  -1.36,  -0.20,  0.34,   -1.13,  -0.88,  2.02,   -1.34]\n",
    "    letterDict[\"GU\"] = [-0.14,  -1.64,  1.50,   0.49,   0.54,   0.34,   -0.27,  -0.31,  0.23,   -0.26]\n",
    "    letterDict[\"GT\"] = [-0.14,  -1.64,  1.50,   0.49,   0.54,   0.34,   -0.27,  -0.31,  0.23,   -0.26]\n",
    "    letterDict[\"UA\"] = [0.09,   0.98,   -0.62,  0.39,   -0.95,  0.34,   1.34,   0.27,   -1.23,  1.08]\n",
    "    letterDict[\"TA\"] = [0.09,   0.98,   -0.62,  0.39,   -0.95,  0.34,   1.34,   0.27,   -1.23,  1.08]\n",
    "    letterDict[\"UC\"] = [1.47,   0.40,   0.14,   -0.94,  1.28,   0.34,   -0.72,  -0.31,  -0.04,  -0.26]\n",
    "    letterDict[\"TC\"] = [1.47,   0.40,   0.14,   -0.94,  1.28,   0.34,   -0.72,  -0.31,  -0.04,  -0.26]\n",
    "    letterDict[\"UG\"] = [-1.87,  0.62,   0.48,   0.33,   0.83,   -0.26,  0.15,   0.82,   -0.17,  0.28]\n",
    "    letterDict[\"TG\"] = [-1.87,  0.62,   0.48,   0.33,   0.83,   -0.26,  0.15,   0.82,   -0.17,  0.28]\n",
    "    letterDict[\"UU\"] = [-0.83,  -0.67,  -1.13,  1.34,   -1.84,  -0.26,  1.72,   0.95,   0.29,   1.35]\n",
    "    letterDict[\"TT\"] = [-0.83,  -0.67,  -1.13,  1.34,   -1.84,  -0.26,  1.72,   0.95,   0.29,   1.35]\n",
    "\n",
    "    AACategoryLen = 10\n",
    "\n",
    "    probMatr = np.zeros((len(shuffled_SEQ), 1, len(shuffled_SEQ[0])-1, AACategoryLen))\n",
    "    sampleNo = 0\n",
    "    for sequence in shuffled_SEQ:\n",
    "        list = []\n",
    "        for a in range(len(sequence)):\n",
    "            try:\n",
    "                b = sequence[a] + sequence[a+1]\n",
    "                list.append(b)\n",
    "            except:\n",
    "                continue\n",
    "        AANo = 0\n",
    "        for AA in list:\n",
    "            try:\n",
    "                if AA in letterDict:\n",
    "                    probMatr[sampleNo][0][AANo] = letterDict[AA]\n",
    "                else:\n",
    "                    continue\n",
    "                AANo += 1\n",
    "            except:\n",
    "                AANo += 1\n",
    "        sampleNo += 1\n",
    "    return probMatr, tag\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ac90130",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################EnsemPseU-master#########################################\n",
    "#############EnsemPseU: Identifying Pseudouridine Site With an Ensemble Approach##################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f0be5a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#二进制单热编码\n",
    "def Binary(sequences):\n",
    "    AA = 'ACGU'\n",
    "    binary_feature = []\n",
    "    for seq in sequences:\n",
    "        binary = []\n",
    "        for aa in seq:\n",
    "            for aa1 in AA:\n",
    "                tag = 1 if aa == aa1 else 0\n",
    "                binary.append(tag)\n",
    "        binary_feature.append(binary)\n",
    "    return binary_feature\n",
    "\n",
    "\n",
    "#理化性质编码\n",
    "def NCP(sequences):\n",
    "    chemical_property = {\n",
    "        'A': [1, 1, 1],\n",
    "        'C': [0, 1, 0],\n",
    "        'G': [1, 0, 0],\n",
    "        'U': [0, 0, 1], }\n",
    "    ncp_feature = []\n",
    "    for seq in sequences:\n",
    "        ncp = []\n",
    "        for aaindex, aa in enumerate(seq):\n",
    "            ncp = ncp + chemical_property.get(aa, [0, 0, 0])\n",
    "        ncp_feature.append(ncp)\n",
    "    return ncp_feature\n",
    "\n",
    "#单核苷酸密度\n",
    "def ND(sequences):\n",
    "    nd_feature = []\n",
    "    for seq in sequences:\n",
    "        nd = []\n",
    "        for aaindex, aa in enumerate(seq):\n",
    "            nd.append(seq[0: aaindex + 1].count(seq[aaindex]) / (aaindex + 1))\n",
    "        nd_feature.append(nd)\n",
    "    return nd_feature\n",
    "\n",
    "\n",
    "#窗口频率编码\n",
    "def ENAC(sequences):\n",
    "    AA = 'ACGU'\n",
    "    enac_feature = []\n",
    "    window = 5\n",
    "    for seq in sequences:\n",
    "        l = len(seq)\n",
    "        enac= []\n",
    "        for i in range(0, l):\n",
    "            if i < l and i + window <= l:\n",
    "                count = Counter(seq[i:i + window])\n",
    "                for key in count:\n",
    "                    count[key] = count[key] / len(seq[i:i + window])\n",
    "                for aa in AA:\n",
    "                    enac.append(count[aa])\n",
    "        enac_feature.append(enac)\n",
    "    return enac_feature\n",
    "\n",
    "\n",
    "#计算单核、双核、三核的频率\n",
    "def Kmer123(sequences):\n",
    "    AA = 'ACGU'\n",
    "    AADict = {}\n",
    "    for i in range(len(AA)):\n",
    "        AADict[AA[i]] = i\n",
    "    kmer123_feature = []\n",
    "\n",
    "    for seq in sequences:\n",
    "        kmer1 = [0] * 4\n",
    "        for j in range(len(seq)):\n",
    "            kmer1[AADict[seq[j]]] = kmer1[AADict[seq[j]]] + 1\n",
    "        if sum(kmer1) != 0:\n",
    "            kmer1 = [i / sum(kmer1) for i in kmer1]\n",
    "\n",
    "        kmer2 = [0] * 16\n",
    "        for j in range(len(seq) - 2 + 1):\n",
    "            kmer2[AADict[seq[j]] * 4 + AADict[seq[j + 1]]] = kmer2[AADict[seq[j]] * 4 + AADict[seq[j + 1]]] + 1\n",
    "        if sum(kmer2) != 0:\n",
    "            kmer2 = [i / sum(kmer2) for i in kmer2]\n",
    "\n",
    "        kmer3 = [0] * 64\n",
    "        for j in range(len(seq) - 3 + 1):\n",
    "            kmer3[AADict[seq[j]] * 16 + AADict[seq[j + 1]] * 4 + AADict[seq[j + 2]]] = kmer3[AADict[seq[j]] * 16 + AADict[seq[j + 1]] * 4 + AADict[seq[j + 2]]] + 1\n",
    "        if sum(kmer3) != 0:\n",
    "            kmer3 = [i / sum(kmer3) for i in kmer3]\n",
    "\n",
    "        kmer = kmer1+kmer2 + kmer3\n",
    "        kmer123_feature.append(kmer)\n",
    "    return kmer123_feature\n",
    "\n",
    "\n",
    "#四核苷酸频率\n",
    "def Kmer4(sequences):\n",
    "    AA = 'ACGU'\n",
    "    AADict = {}\n",
    "    for i in range(len(AA)):\n",
    "        AADict[AA[i]] = i\n",
    "    kmer4 = []\n",
    "    for seq in sequences:\n",
    "        tmpCode = [0] * 256\n",
    "        for j in range(len(seq) - 4 + 1):\n",
    "            tmpCode[AADict[seq[j]] * 64 + AADict[seq[j + 1]] * 16 + AADict[seq[j + 2]] * 4 + AADict[seq[j + 3]]] = tmpCode[AADict[seq[j]] * 64 + AADict[seq[j + 1]] * 16 + AADict[seq[j + 2]] * 4 + AADict[seq[j + 3]]] + 1\n",
    "        if sum(tmpCode) != 0:\n",
    "            tmpCode = [i / sum(tmpCode) for i in tmpCode]\n",
    "        kmer4.append(tmpCode)\n",
    "    return kmer4\n",
    "\n",
    "\n",
    "#五核苷酸频率\n",
    "def Kmer5(sequences):\n",
    "    AA = 'ACGU'\n",
    "    kmer5 = []\n",
    "    AADict = {}\n",
    "    for i in range(len(AA)):\n",
    "        AADict[AA[i]] = i\n",
    "    for seq in sequences:\n",
    "        tmpCode = [0] * 1024\n",
    "        for j in range(len(seq) - 5 + 1):\n",
    "            tmpCode[AADict[seq[j]] * 256 + AADict[seq[j + 1]] * 64 + AADict[seq[j + 2]] * 16 + AADict[seq[j + 3]] * 4 +\n",
    "                    AADict[seq[j + 4]]] = tmpCode[\n",
    "                                              AADict[seq[j]] * 256 + AADict[seq[j + 1]] * 64 + AADict[seq[j + 2]] * 16 +\n",
    "                                              AADict[seq[j + 3]] * 4 + AADict[seq[j + 4]]] + 1\n",
    "        if sum(tmpCode) != 0:\n",
    "            tmpCode = [i / sum(tmpCode) for i in tmpCode]\n",
    "        kmer5.append(tmpCode)\n",
    "\n",
    "    return kmer5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a9434a",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############i4mC-Deep-main##########################################################\n",
    "###########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3d10e25",
   "metadata": {},
   "outputs": [],
   "source": [
    "#nucleotide chemical properties and frequency encoding\n",
    "#单核苷酸理化性质+频率编码\n",
    "def cal(c,cb,i):\n",
    "    bases ={'A':[1,1,1], 'C':[0,1,0], 'G':[1,0,0,], 'T':[0,0,1]}\n",
    "    p=[]\n",
    "    p=bases[c]\n",
    "    p.append(np.round(cb/float(i+1),2))\n",
    "    return(p)\n",
    "def calculate(s):\n",
    "    p=f=list()\n",
    "    cba=cbc=cbt=cbg=0\n",
    "    for i,c in enumerate(s):\n",
    "        if c=='A':\n",
    "            cba+=1\n",
    "            p=cal(c,cba,i)\n",
    "        elif c=='T':\n",
    "            cbt+=1\n",
    "            p=cal(c,cbt,i)\n",
    "        elif c=='C':\n",
    "            cbc+=1\n",
    "            p=cal(c,cbc,i)\n",
    "        elif c=='G':\n",
    "            cbg+=1\n",
    "            p=cal(c,cbg,i)\n",
    "        else:\n",
    "            p=[0,0,0,0]\n",
    "        f.append(p)\n",
    "    return(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e439750",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################PseUdeep-master########################################################3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e5f0ad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############R5hmCFDV-main############################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adcaa7cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#########NV###############################\n",
    "#这三个用来计算自然向量\n",
    "###convert every A, C, F, and T to number 0, 1, 2, and 3###\n",
    "def tran_digital(ch):\n",
    "    if ch == 'A':\n",
    "        return 0\n",
    "    elif ch == 'C':\n",
    "        return 1\n",
    "    elif ch == 'G':\n",
    "        return 2\n",
    "    else:\n",
    "        return 3\n",
    "\n",
    "\n",
    "##convert every sequence to decimal number. Foe example, if window size is 3, AAAAAA-TTTTTT could be represent 0-4095.###\n",
    "def cal_label(feature):\n",
    "    index = 0\n",
    "    for i in range(0, len(feature)):\n",
    "        index = 4 * index + feature[i]\n",
    "    return index\n",
    "\n",
    "\n",
    "def NV(file):\n",
    "    win_size = 1 #window \n",
    "    filename = 'NV_data'\n",
    "    out = open(filename + \".txt\", \"w\" ) #outputFile\n",
    "\n",
    "    seq = []\n",
    "    for line in file:\n",
    "        if line[0] == '>':  # pass annotation line\n",
    "            continue\n",
    "        line = line.upper()  # convert every lower nucleotide to upper nucleotide\n",
    "        line = line.strip('\\n')  \n",
    "        number = [0] * len(line)\n",
    "\n",
    "        # convert every A, C, F, and T to number 0, 1, 2, and 3\n",
    "        for i in range(0, len(line)):\n",
    "            number[i] = tran_digital(line[i])\n",
    "    \n",
    "        # calculate frequency\n",
    "        nk_frequence = [0] * (4 ** win_size)\n",
    "        tk_position = [0] * (4 ** win_size)\n",
    "        u = [0] * (4 ** win_size)\n",
    "        for i in range(0, len(number) - win_size + 1):\n",
    "            id = cal_label(number[i:i + win_size])\n",
    "            nk_frequence[id] += 1\n",
    "            tk_position[id] +=i+1\n",
    "        for i in range(0,4**win_size):\n",
    "            if nk_frequence[i] == 0 :\n",
    "                u[i] = 0\n",
    "            else:\n",
    "                u[i] = tk_position[i]/nk_frequence[i]\n",
    "    \n",
    "\t    \n",
    "        dk_secondery = [0] * (4 ** win_size)\n",
    "        for i in range(0, len(number) - win_size + 1):\n",
    "            id = cal_label(number[i:i + win_size])\n",
    "            dk_secondery[id] += ((i+1-u[id])**2)/(nk_frequence[id]*len(number))\n",
    "\t    \t\n",
    "        out.write(\"\\t\")\n",
    "        for i in range(0,len(dk_secondery)):\n",
    "            out.write(\"%d:%.4f\\t\" % (1+3*i, nk_frequence[i]))\n",
    "            out.write(\"%d:%.4f\\t\" % (2+3*i, u[i]))\n",
    "            out.write(\"%d:%.4f\\t\" % (3+3*i, dk_secondery[i]))\n",
    "        out.write('\\n')\n",
    "    #out.close()\n",
    "    return out\n",
    "\n",
    "\n",
    "######PCP########\n",
    "for line in fh:#get the fasta sequence\n",
    "    if line.startswith('>'):\n",
    "        pass\n",
    "    else:\n",
    "        seq.append(line.replace('\\n','').replace('\\r',''))\n",
    "fh.close()  #获取序列\n",
    "\n",
    "\n",
    "data=pd.read_csv(physical_chemical_properties_path,header=None,index_col=None)#read the phisical chemichy proporties\n",
    "prop_key=data.values[:,0]  #读取理化性质的key，第0列\n",
    "\n",
    "if fill_NA==\"1\":\n",
    "    prop_key[21]='NA'  #是否添加“NA”\n",
    "#print(prop_key)\n",
    "prop_data=data.values[:,1:]  #读取理化性质的values值，从1开始往后所有列\n",
    "prop_data=np.matrix(prop_data)\n",
    "DNC_value=np.array(prop_data).T  #DNC转置一下\n",
    "DNC_value_scale=[[]]*len(DNC_value)\n",
    "for i in list(range(len(DNC_value))):\n",
    "    average_=sum(DNC_value[i]*1.0/len(DNC_value[i]))\n",
    "    std_=np.std(DNC_value[i],ddof=1)\n",
    "    DNC_value_scale[i]=[round((e-average_)/std_,2) for e in DNC_value[i]]\n",
    "prop_data_transformed=list(zip(*DNC_value_scale))\n",
    "# prop_data_transformed=StandardScaler().fit_transform(prop_data)\n",
    "prop_len=len(prop_data_transformed[0])\n",
    "\n",
    "whole_m6a_seq=seq\n",
    "i=0\n",
    "phisical_chemichy_len=len(prop_data_transformed)#the length of properties\n",
    "sequence_line_len=len(seq[0])#the length of one sequence\n",
    "LAMDA=5\n",
    "finally_result=[]#used to save the fanal result\n",
    "for one_m6a_sequence_line in whole_m6a_seq:\n",
    "    one_sequence_value=[[]]*(sequence_line_len-1)\n",
    "    PC_m=[0.0]*prop_len\n",
    "    PC_m=np.array(PC_m)\n",
    "    for one_sequence_index in range(sequence_line_len-1):\n",
    "        for prop_index in list(range(len(prop_key))):\n",
    "            if one_m6a_sequence_line[one_sequence_index:one_sequence_index+2]==prop_key[prop_index]:\n",
    "                one_sequence_value[one_sequence_index]=prop_data_transformed[prop_index]\n",
    "        PC_m+=np.array(one_sequence_value[one_sequence_index])\n",
    "    PC_m=PC_m/(sequence_line_len-1)\n",
    "    auto_value=[]\n",
    "    for LAMDA_index in list(range(1,LAMDA+1)):\n",
    "        temp = [0.0] * prop_len\n",
    "        temp=np.array(temp)\n",
    "        for auto_index in list(range(1,sequence_line_len-LAMDA_index)):\n",
    "            temp=temp+(np.array(one_sequence_value[auto_index-1])-PC_m)*(np.array(one_sequence_value[auto_index+LAMDA_index-1])-PC_m)\n",
    "            temp=[round(e,8) for e in temp.astype(float)]\n",
    "        x=[round(e/(sequence_line_len-LAMDA_index-1),8) for e in temp]\n",
    "        auto_value.extend([round(e,8) for e in x])\n",
    "    for LAMDA_index in list(range(1, LAMDA + 1)):\n",
    "        for i in list(range(1,prop_len+1)):\n",
    "            for j in list(range(1,prop_len+1)):\n",
    "                temp2=0.0\n",
    "                if i != j:\n",
    "                    for auto_index in list(range(1, sequence_line_len - LAMDA_index)):\n",
    "                            temp2+=(one_sequence_value[auto_index-1][i-1]-PC_m[i-1])*(one_sequence_value[auto_index+LAMDA_index-1][j-1]-PC_m[j-1])\n",
    "                    auto_value.append(round(temp2/((sequence_line_len-1)-LAMDA_index),8))\n",
    "\n",
    "    finally_result.append(auto_value)\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "########PseDNC#######\n",
    "phisical_chemical_proporties=pd.read_csv(path+propertyname,header=None,index_col=None)\n",
    "m6a_sequence=open(r'E:\\论文\\5hmcdeep\\Data\\Data2\\Dataset.txt', 'r')\n",
    "DNC_key=phisical_chemical_proporties.values[:,0]\n",
    "#print (DNC_key)\n",
    "#print (len(DNC_key))\n",
    "\n",
    "if fill_NA==\"1\":\n",
    "    DNC_key[21]='NA'\n",
    "# DNC_key=np.array(['AA','AC','AG','AU','CA','CC','CG','CU','GA','GC','GG','GU','UA','UC','UG','UU'])\n",
    "DNC_value=phisical_chemical_proporties.values[:,1:]\n",
    "DNC_value=np.array(DNC_value).T\n",
    "DNC_value_scale=[[]]*len(DNC_value)\n",
    "#print (len(DNC_value))\n",
    "for i in range(len(DNC_value)):\n",
    "    average_=sum(DNC_value[i]*1.0/len(DNC_value[i]))\n",
    "    std_=np.std(DNC_value[i],ddof=1)\n",
    "    DNC_value_scale[i]=[round((e-average_)/std_,2) for e in DNC_value[i]]\n",
    "    #print (DNC_value_scale)\n",
    "DNC_value_scale=list(zip(*DNC_value_scale))\n",
    "#print DNC_value_scale\n",
    "\n",
    "\n",
    "\n",
    "DNC_len=len(DNC_value_scale)\n",
    "#print (DNC_len)\n",
    "m6aseq=[]\n",
    "for line in m6a_sequence:\n",
    "    if line.startswith('>'):\n",
    "        pass\n",
    "    elif line == '\\n':\n",
    "        line = line.strip(\"\\n\")\n",
    "    else:\n",
    "        m6aseq.append(line.replace('\\n','').replace(\"\\r\",''))\n",
    "w=0.9\n",
    "Lamda=6\n",
    "result_value=[]\n",
    "m6a_len=len(m6aseq[0])\n",
    "#print m6a_len\n",
    "m6a_num=len(m6aseq)\n",
    "for m6a_line_index in range(m6a_num):\n",
    "    frequency=[0]*len(DNC_key)\n",
    "    #print len(frequency)\n",
    "    m6a_DNC_value=[[]]*(m6a_len-1)\n",
    "    #print m6a_DNC_value\n",
    "    for m6a_line_doublechar_index in range(m6a_len):\n",
    "        for DNC_index in range(len(DNC_key)):\n",
    "            if m6aseq[m6a_line_index][m6a_line_doublechar_index:m6a_line_doublechar_index+2]==DNC_key[DNC_index]:\n",
    "                #print m6aseq[2][0:2]\n",
    "                m6a_DNC_value[m6a_line_doublechar_index]=DNC_value_scale[DNC_index]\n",
    "                frequency[DNC_index]+=1\n",
    "    #print m6a_DNC_value\n",
    "    \n",
    "    frequency=[e/float(sum(frequency)) for e in frequency]\n",
    "    p=sum((frequency))\n",
    "    #print p\n",
    "    #frequency=np.array(frequency)/float(sum(frequency))#(m6a_len-1)\n",
    "    one_line_value_with = 0.0\n",
    "    sita = [0] * Lamda\n",
    "    #print len(sita)\n",
    "    for lambda_index in range(1, Lamda + 1):\n",
    "        one_line_value_without_ = 0.0\n",
    "        for m6a_sequence_value_index in range(1, m6a_len - lambda_index):\n",
    "            temp = list(map(lambda x,y : round((x - y) ** 2,8), list(np.array(m6a_DNC_value[m6a_sequence_value_index - 1])),list(np.array(m6a_DNC_value[m6a_sequence_value_index - 1 + lambda_index]))))\n",
    "            #map(lambda x, y: x + y, [1, 3, 5, 7, 9], [2, 4, 6, 8, 10])\n",
    "            temp_value = round(sum(temp) * 1.0 / DNC_len,8)\n",
    "            one_line_value_without_ += temp_value\n",
    "        one_line_value_without_ = round(one_line_value_without_ / (m6a_len - lambda_index-1),8)\n",
    "        sita[lambda_index - 1] = one_line_value_without_\n",
    "        one_line_value_with += one_line_value_without_\n",
    "    dim = [0] * (len(DNC_key) + Lamda)\n",
    "    #print len(dim)\n",
    "    for index in range(1, len(DNC_key) + Lamda+1):\n",
    "        if index <= len(DNC_key):\n",
    "            dim[index - 1] = frequency[index - 1] / (1.0 + w * one_line_value_with)\n",
    "        else:\n",
    "            dim[index - 1] = w * sita[index - len(DNC_key)-1] / (1.0 + w * one_line_value_with)\n",
    "        dim[index-1]=round(dim[index-1],8)\n",
    "    result_value.append(dim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "916274e8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-29T08:02:03.901282Z",
     "start_time": "2022-11-29T08:02:03.889721Z"
    }
   },
   "outputs": [],
   "source": [
    "#核苷酸化学性质+单核苷酸频率，返回的shape为batch_Size*input_length*4\n",
    "def NCPA(sequences):\n",
    "    chemical_property = {\n",
    "        'A': [1, 1, 1],\n",
    "        'T': [0, 1, 0],\n",
    "        'G': [1, 0, 0],\n",
    "        'C': [0, 0, 1]\n",
    "    }\n",
    "    ncp_feature = []\n",
    "    for seq in sequences:\n",
    "        ncp = []\n",
    "        for aaindex, aa in enumerate(seq):\n",
    "            ncp.append(chemical_property.get(aa, [0, 0, 0]))\n",
    "        ncp=np.append(ncp,calculate(seq).reshape(-1,1),axis=1)\n",
    "        ncp_feature.append(ncp)\n",
    "    return np.array(ncp_feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "69a9b07a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-29T08:02:04.462850Z",
     "start_time": "2022-11-29T08:02:04.459319Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "88c555eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-29T08:02:24.753376Z",
     "start_time": "2022-11-29T08:02:24.745746Z"
    }
   },
   "outputs": [],
   "source": [
    "def calculate(sequence):\n",
    "\n",
    "    X = []\n",
    "    dictNum = {'A' : 0, 'T' : 0, 'C' : 0, 'G' : 0, 'N': 0};\n",
    "    for i in range(len(sequence)):\n",
    "        if sequence[i] in dictNum.keys():\n",
    "            dictNum[sequence[i]] += 1;\n",
    "            X.append(dictNum[sequence[i]] / float(i + 1));\n",
    "    return np.array(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "85193dc5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-11-29T08:02:38.190121Z",
     "start_time": "2022-11-29T08:02:38.181710Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[[1.        , 1.        , 1.        , 1.        ],\n",
       "        [0.        , 1.        , 0.        , 0.5       ],\n",
       "        [0.        , 0.        , 1.        , 0.33333333],\n",
       "        [1.        , 0.        , 0.        , 0.25      ],\n",
       "        [1.        , 1.        , 1.        , 0.4       ],\n",
       "        [0.        , 1.        , 0.        , 0.33333333],\n",
       "        [0.        , 0.        , 1.        , 0.28571429],\n",
       "        [1.        , 0.        , 0.        , 0.25      ],\n",
       "        [0.        , 1.        , 0.        , 0.33333333]],\n",
       "\n",
       "       [[0.        , 1.        , 0.        , 1.        ],\n",
       "        [1.        , 1.        , 1.        , 0.5       ],\n",
       "        [1.        , 0.        , 0.        , 0.33333333],\n",
       "        [0.        , 0.        , 1.        , 0.25      ],\n",
       "        [0.        , 1.        , 0.        , 0.4       ],\n",
       "        [1.        , 1.        , 1.        , 0.33333333],\n",
       "        [1.        , 0.        , 0.        , 0.28571429],\n",
       "        [0.        , 1.        , 0.        , 0.375     ],\n",
       "        [1.        , 1.        , 1.        , 0.33333333]]])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq=[\"ATCGATCGT\",\"TAGCTAGTA\"]\n",
    "NCPA(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "40f409be",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-12-05T13:59:33.182665Z",
     "start_time": "2022-12-05T13:59:33.166737Z"
    }
   },
   "outputs": [],
   "source": [
    "from gensim.models import word2vec\n",
    "\n",
    "def emb_seqs(sequences, features=100, num = 4):\n",
    "    w2v_model = word2vec.Word2Vec.load(\"dna_w2v_100.pt\")\n",
    "    seqs_emb = []\n",
    "    for seq in sequences:\n",
    "        seq_emb = []\n",
    "        for i in range(len(seq) - num + 1):\n",
    "            try:\n",
    "                seq_emb.append(np.array(w2v_model.wv[seq[i:i+num]]))\n",
    "            except:\n",
    "                seq_emb.append(np.array(np.zeros([features])))\n",
    "        seqs_emb.append(seq_emb)\n",
    "    seqs_emb = np.array(seqs_emb).reshape(len(seqs_emb),-1,features)\n",
    "    return seqs_emb #词向量编码"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "486e452d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-12-09T15:53:45.455576Z",
     "start_time": "2022-12-09T15:53:45.390958Z"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    " #########################PSTNPSS########################################################\n",
    "def CalculateMatrix_PSTNPss(data, order):\n",
    "    matrix = np.zeros((len(data[0]) - 2, 64))\n",
    "    for i in range(len(data[0]) - 2):  # position\n",
    "        for j in range(len(data)):\n",
    "            if re.search('-', data[j][i:i + 3]):\n",
    "                pass\n",
    "            else:\n",
    "                matrix[i][order[data[j][i:i + 3]]] += 1   #第几个位置，以及对应核酸种类\n",
    "    return matrix\n",
    "\n",
    "\n",
    "def PSTNPss(inf, sequence_all, cell_line):\n",
    "    positive = []  # load\n",
    "    negative = []  # load\n",
    "    for i in range(len(sequence_all)):\n",
    "        if cell_line == 'IMR90/':\n",
    "            positive.append(sequence_all[i])\n",
    "        else:\n",
    "            if i < len(sequence_all) / 2:\n",
    "                positive.append(sequence_all[i])\n",
    "            else:\n",
    "                negative.append(sequence_all[i])  #半阳半阴\n",
    "\n",
    "    encodings = []  #空编码\n",
    "    header = ['#', 'label']  #header\n",
    "    fastas = []   #fastas\n",
    "    f = open(inf, 'r')\n",
    "    for i in f.readlines():\n",
    "        list_line = list(i)  #这一行变成list\n",
    "        if list_line[0] == '>':\n",
    "            continue\n",
    "        else:\n",
    "            if i[-1] == '\\n':  #若最后是换行符\n",
    "                sequence = i[0:-1]  #最后不取\n",
    "            else:\n",
    "                sequence = i  #sequence为序列\n",
    "            fastas.append(sequence)  #添加到fastas\n",
    "    for pos in range(len(fastas[0]) - 2):  #长度-2\n",
    "        header.append('Pos.%d' % (pos + 1)) #header添加\n",
    "    encodings.append(header)  #encodings添加\n",
    "\n",
    "    '''for i in range(int(len(fastas)/2)):\n",
    "        positive.append(fastas[i])\n",
    "        negative.append(fastas[i+97])\n",
    "    print(len(positive))'''\n",
    "\n",
    "    nucleotides = ['A', 'C', 'G', 'T']  #定义核酸字典\n",
    "    trinucleotides = [n1 + n2 + n3 for n1 in nucleotides for n2 in nucleotides for n3 in nucleotides]  #三核苷酸列表\n",
    "    order = {}\n",
    "    for i in range(len(trinucleotides)):\n",
    "        order[trinucleotides[i]] = i  #变成字典\n",
    "    matrix_po = CalculateMatrix_PSTNPss(positive, order)\n",
    "    if cell_line != 'IMR90/':\n",
    "        matrix_ne = CalculateMatrix_PSTNPss(negative, order)\n",
    "\n",
    "    positive_number = len(positive)\n",
    "    if cell_line != 'IMR90/':\n",
    "        negative_number = len(negative)\n",
    "    count = 0\n",
    "    for i in fastas:\n",
    "        count += 1\n",
    "        sequence = i\n",
    "        code = []\n",
    "        if cell_line == 'IMR90/':\n",
    "            for j in range(len(sequence) - 2):\n",
    "                if re.search('-', sequence[j: j + 3]):\n",
    "                    code.append(0)\n",
    "                else:\n",
    "                    p_num = positive_number\n",
    "                    po_number = matrix_po[j][order[sequence[j: j + 3]]]\n",
    "                    if count < int(len(fastas)) and po_number > 0:\n",
    "                        po_number -= 1\n",
    "                        p_num -= 1\n",
    "                    code.append(po_number/p_num)\n",
    "                    # print(sequence[j: j+3], order[sequence[j: j+3]], po_number, p_num, ne_number, n_num)\n",
    "            encodings.append(code)\n",
    "        else:\n",
    "            for j in range(len(sequence) - 2):\n",
    "                if re.search('-', sequence[j: j + 3]):\n",
    "                    code.append(0)\n",
    "                else:\n",
    "                    p_num, n_num = positive_number, negative_number\n",
    "                    po_number = matrix_po[j][order[sequence[j: j + 3]]]\n",
    "                    ne_number = matrix_ne[j][order[sequence[j: j + 3]]]\n",
    "                    if count < int(len(fastas) / 2) and po_number > 0:\n",
    "                        po_number -= 1\n",
    "                        p_num -= 1\n",
    "                    if count >= int(len(fastas) / 2) and ne_number > 0:\n",
    "                        ne_number -= 1\n",
    "                        n_num -= 1\n",
    "                    code.append(po_number / p_num - ne_number / n_num)\n",
    "                    # print(sequence[j: j+3], order[sequence[j: j+3]], po_number, p_num, ne_number, n_num)\n",
    "            encodings.append(code)\n",
    "    return encodings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c9944cfc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-12-09T15:54:10.550436Z",
     "start_time": "2022-12-09T15:54:10.533424Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = [1,2,3,4]\n",
    "s[0:-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502297d2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:BERT]",
   "language": "python",
   "name": "conda-env-BERT-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
